File: examples/app-redux-android-todolist-kotlin/src/main/java/com/redux/ActionCreator.java
package com.redux;

import rx.Subscriber;

import javax.inject.Inject;
import java.util.List;

public class ActionCreator {
    private final Store<AppAction, AppState> store;
    private final Operations operations;

    @Inject
    public ActionCreator(Store<AppAction, AppState> store, Operations operations) {
        this.store = store;
        this.operations = operations;
    }

    public rx.Subscription fetch() {
        store.dispatch(new AppAction.Fetching(true));
        return operations
                .fetch()
                .subscribe(new Subscriber<List<Todo>>() {
                    @Override public void onCompleted() {
                        store.dispatch(new AppAction.Fetching(false));
                    }

                    @Override public void onError(Throwable e) {
                        store.dispatch(new AppAction.Fetching(false));
                    }

                    @Override public void onNext(List<Todo> todos) {
                        // TODO : in a real wworld situation dispatch all at once.
                        for (Todo todo : todos) {
                            store.dispatch(new AppAction.Add(todo.getText(), todo.getIsCompleted()));
                        }
                    }
                });
    }

    public void add(String text) {
        store.dispatch(new AppAction.Add(text, false));
    }

    public void add(String text, boolean isCompleted) {
        store.dispatch(new AppAction.Add(text, isCompleted));
    }

    public void delete(int id) {
        store.dispatch(new AppAction.Delete(id));
    }

    public void complete(int id, boolean isCompleted) {
        store.dispatch(new AppAction.Complete(id, isCompleted));
    }

    public void completeAll(boolean isCompleted) {
        store.dispatch(new AppAction.CompleteAll(isCompleted));
    }

    public void clearCompleted() {
        store.dispatch(new AppAction.ClearCompleted());
    }

}

File: examples/app-redux-android-todolist-kotlin/src/main/kotlin/com/redux/ActionCreator.kt
package com.redux

import javax.inject.Inject

public class ActionCreator @Inject constructor(
        private val store: Store<TodoListAction, TodoListState>,
        private val operations: Operations) {

    public fun fetch(): rx.Subscription {
        store.dispatch(TodoListAction.Fetching(true))
        return operations
                .fetch()
                .doOnNext { todos -> todos.forEach { store.dispatch(TodoListAction.Add(it.text, it.isCompleted)) } }
                .doOnCompleted { store.dispatch(TodoListAction.Fetching(false)) }
                .doOnError { store.dispatch(TodoListAction.Fetching(false)) }
                .subscribe()
    }

    public fun add(text: String) = store.dispatch(TodoListAction.Add(text, false))

    public fun add(text: String, isCompleted: Boolean) = store.dispatch(TodoListAction.Add(text, isCompleted))

    public fun delete(id: Int) = store.dispatch(TodoListAction.Delete(id))

    public fun complete(id: Int, isCompleted: Boolean) = store.dispatch(TodoListAction.Complete(id, isCompleted))

    public fun completeAll(isCompleted: Boolean) = store.dispatch(TodoListAction.CompleteAll(isCompleted))

    public fun clearCompleted() = store.dispatch(TodoListAction.ClearCompleted)

}
